storage/postgres/log_storage.go

```diff
@@ -0,0 +1,1011 @@
+		}
+
+		leaves = append(leaves, leaf)
+		dq = append(dq, dqInfo)
+	}
+
+	if rows.Err() != nil {
+		return nil, rows.Err()
+	}
+	label := labelForTX(t)
+	selectDuration := time.Since(start)
+	observe(dequeueSelectLatency, selectDuration, label)
+
+	// The convention is that if leaf processing succeeds (by committing this tx)
+	// then the unsequenced entries for them are removed
+	if len(leaves) > 0 {
+		err = t.removeSequencedLeaves(ctx, dq)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	totalDuration := time.Since(start)
+	removeDuration := totalDuration - selectDuration
+	observe(dequeueRemoveLatency, removeDuration, label)
+	observe(dequeueLatency, totalDuration, label)
+	dequeuedCounter.Add(float64(len(leaves)), label)
+
+	return leaves, nil
+}
+
+// sortLeavesForInsert returns a slice containing the passed in leaves sorted
+// by LeafIdentityHash, and paired with their original positions.
+// QueueLeaves and AddSequencedLeaves use this to make the order that LeafData
+// row locks are acquired deterministic and reduce the chance of deadlocks.
+func sortLeavesForInsert(leaves []*trillian.LogLeaf) []leafAndPosition {
+	ordLeaves := make([]leafAndPosition, len(leaves))
+	for i, leaf := range leaves {
+		ordLeaves[i] = leafAndPosition{leaf: leaf, idx: i}
+	}
+	sort.Sort(byLeafIdentityHashWithPosition(ordLeaves))
+	return ordLeaves
+}
+
+func (t *logTreeTX) QueueLeaves(ctx context.Context, leaves []*trillian.LogLeaf, queueTimestamp time.Time) ([]*trillian.LogLeaf, error) {
+	// Don't accept batches if any of the leaves are invalid.
+	for _, leaf := range leaves {
+		if len(leaf.LeafIdentityHash) != t.hashSizeBytes {
+			return nil, fmt.Errorf("queued leaf must have a leaf ID hash of length %d", t.hashSizeBytes)
+		}
+		var err error
+		leaf.QueueTimestamp, err = ptypes.TimestampProto(queueTimestamp)
+		if err != nil {
+			return nil, fmt.Errorf("got invalid queue timestamp: %v", err)
+		}
+	}
+	start := time.Now()
+	label := labelForTX(t)
+
+	ordLeaves := sortLeavesForInsert(leaves)
+	existingCount := 0
+	existingLeaves := make([]*trillian.LogLeaf, len(leaves))
+
+	for _, ol := range ordLeaves {
+		i, leaf := ol.idx, ol.leaf
+
+		leafStart := time.Now()
+		qTimestamp, err := ptypes.Timestamp(leaf.QueueTimestamp)
+		if err != nil {
+			return nil, fmt.Errorf("got invalid queue timestamp: %v", err)
+		}
+		dupCheckRow, err := t.tx.QueryContext(ctx, insertLeafDataSQL, t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData, qTimestamp.UnixNano())
+		insertDuration := time.Since(leafStart)
+		observe(queueInsertLeafLatency, insertDuration, label)
+		var resultData = true
+		dupCheckRow.Scan(&resultData)
+		dupCheckRow.Close()
+		//if isDuplicateErr(err) {
+		if !resultData {
+			// Remember the duplicate leaf, using the requested leaf for now.
+			existingLeaves[i] = leaf
+			existingCount++
+			queuedDupCounter.Inc(label)
+			glog.Warningf("Found duplicate %v %v", t.treeID, leaf)
```

Comment:
You probably don't want this as a straight up `Warning`, perhaps you could make it a `glog.V(1).Warning(...`?